File: bot/src/main/java/com/vpnsupport/llm/OpenAiClient.java
package com.vpnsupport.llm;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.vpnsupport.bot.ChatHistoryService;
import com.vpnsupport.bot.LlmTokenUsage;
import com.vpnsupport.bot.LlmTokenUsageRepository;
import com.vpnsupport.config.OpenAiProperties;
import com.vpnsupport.rag.FaqEmbeddingService;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;
import reactor.netty.http.client.HttpClient;

import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Component
@ConditionalOnProperty(name = "llm.provider", havingValue = "openai")
public class OpenAiClient extends AbstractLlmClient {

    private static final String ROLE_KEY = "role";
    private static final String CONTENT_KEY = "content";
    private static final String CALL_ID_KEY = "call_id";


    private final WebClient webClient;
    private final String model;
    private final Double temperature;
    /**
     * Built once at construction: {@link McpRouter} already has the tool list by
     * the time this bean is created, so the lazy double-checked locking this
     * replaced never had anything to defer.
     */
    private final List<Map<String, Object>> toolDefinitions;

    public OpenAiClient(OpenAiProperties properties, ObjectMapper objectMapper,
                        McpRouter mcpRouter, ChatHistoryService chatHistoryService,
                        FaqEmbeddingService faqEmbeddingService,
                        LlmTokenUsageRepository tokenUsageRepository) {
        super(objectMapper, mcpRouter, chatHistoryService, faqEmbeddingService, tokenUsageRepository);
        String apiKey = properties.getApiKey();
        if (apiKey == null || apiKey.isBlank()) {
            throw new IllegalArgumentException("OpenAI API key must not be null or blank");
        }
        this.model = properties.getModel();
        this.temperature = properties.getTemperature();
        this.webClient = WebClient.builder()
                .baseUrl(properties.getBaseUrl())
                .defaultHeader("Authorization", "Bearer " + apiKey)
                .defaultHeader("Content-Type", "application/json")
                .clientConnector(new org.springframework.http.client.reactive.ReactorClientHttpConnector(
                        HttpClient.create().responseTimeout(Duration.ofSeconds(60))))
                .build();
        this.toolDefinitions = buildToolDefinitions();
    }

    @Override
    public boolean supportsImages() {
        return true;
    }

    @Override
    @SuppressWarnings("unused")
    protected List<Map<String, Object>> buildInitialConversation(
            String userMessage, long telegramUserId, String faqContext,
            String base64Image, String mimeType) {
        List<Map<String, Object>> messages = new ArrayList<>();
        messages.add(Map.of(ROLE_KEY, "system", CONTENT_KEY, SupportPrompt.SYSTEM));

        String dynamicContext = "Telegram ID: " + telegramUserId;
        if (faqContext != null && !faqContext.isEmpty()) {
            dynamicContext += "\n\n" + faqContext;
        }
        messages.add(Map.of(ROLE_KEY, "system", CONTENT_KEY, dynamicContext));

        messages.addAll(chatHistoryService.getHistory(telegramUserId));

        if (base64Image != null && !base64Image.isEmpty()) {
            List<Object> parts = new ArrayList<>();
            if (userMessage != null && !userMessage.isBlank()) {
                parts.add(Map.of("type", "input_text", "text", userMessage));
            }
            String dataUri = "data:" + (mimeType != null ? mimeType : "image/jpeg") + ";base64," + base64Image;
            parts.add(Map.of(
                    "type", "input_image",
                    "image_url", dataUri
            ));
            messages.add(Map.of(ROLE_KEY, "user", CONTENT_KEY, parts));
        } else {
            messages.add(Map.of(ROLE_KEY, "user", CONTENT_KEY, userMessage));
        }

        return messages;
    }

    @Override
    @SuppressWarnings("unused")
    protected String callApi(List<Map<String, Object>> conversation, String faqContext, long telegramUserId) {
        ObjectNode requestBody = objectMapper.createObjectNode();
        requestBody.put("model", model);

        ArrayNode inputArray = requestBody.putArray("input");
        for (Map<String, Object> item : conversation) {
            inputArray.add(objectMapper.valueToTree(item));
        }

        List<Map<String, Object>> tools = toolDefinitions;
        if (!tools.isEmpty()) {
            ArrayNode toolsArray = requestBody.putArray("tools");
            for (Map<String, Object> tool : tools) {
                toolsArray.add(objectMapper.valueToTree(tool));
            }
            requestBody.put("tool_choice", "auto");
            ObjectNode reasoning = requestBody.putObject("reasoning");
            reasoning.put("effort", "none");
        }

        if (temperature != null) {
            requestBody.put("temperature", temperature);
        }

        log.debug("OpenAI Responses API request ({} tools available)", toolDefinitions.size());

        String rawResponse = webClient.post()
                .uri("/responses")
                .bodyValue(requestBody)
                .retrieve()
                .onStatus(status -> status.is4xxClientError() || status.is5xxServerError(),
                        clientResponse -> clientResponse.bodyToMono(String.class)
                                .flatMap(err -> {
                                    if (clientResponse.statusCode().value() == 401) {
                                        return Mono.error(new RuntimeException(
                                                "OpenAI API error (model=" + model + "): "
                                                + clientResponse.statusCode() + " - " + err
                                                + " | Проверьте OPENAI_API_KEY и OPENAI_MODEL в .env"));
                                    }
                                    return Mono.error(new RuntimeException(
                                            "OpenAI API error (model=" + model + "): "
                                            + clientResponse.statusCode() + " - " + err));
                                }))
                .bodyToMono(String.class)
                .block();

        if (rawResponse == null) {
            throw new LlmProcessingException("OpenAI Responses API returned null/empty body",
                    "Не удалось получить ответ от модели. Попробуйте позже.");
        }

        return rawResponse;
    }

    @Override
    protected LlmResponse parseResponse(String rawResponse) {
        try {
            JsonNode jsonResponse = objectMapper.readTree(rawResponse);
            JsonNode output = jsonResponse.get("output");
            if (output == null || !output.isArray()) {
                log.error("No output array in OpenAI Responses API response: {}", rawResponse);
                throw new LlmProcessingException("Empty output",
                        "Не удалось получить ответ от модели. Попробуйте позже.");
            }

            StringBuilder text = new StringBuilder();
            List<LlmResponse.ToolCall> toolCalls = new ArrayList<>();

            for (JsonNode item : output) {
                String type = item.has("type") ? item.get("type").asText() : "";
                if ("function_call".equals(type)) {
                    toolCalls.add(parseToolCall(item));
                } else if ("message".equals(type)) {
                    appendMessageText(item, text);
                }
            }

            if (text.isEmpty() && toolCalls.isEmpty()) {
                log.warn("No text or tool calls in OpenAI response. output={}", rawResponse);
                throw new LlmProcessingException("Empty response",
                        "Модель не вернула ответа. Попробуйте переформулировать вопрос.");
            }

            return new LlmResponse(text.toString(), toolCalls);
        } catch (LlmProcessingException e) {
            throw e;
        } catch (Exception e) {
            log.error("Failed to parse OpenAI response: {} — response: {}", e.getMessage(), rawResponse);
            throw new LlmProcessingException("Parse error: " + e.getMessage(),
                    "Ошибка обработки ответа модели.");
        }
    }

    private LlmResponse.ToolCall parseToolCall(JsonNode item)
            throws com.fasterxml.jackson.core.JsonProcessingException {
        String fnArgsStr = item.get("arguments").asText();
        @SuppressWarnings("unchecked")
        Map<String, Object> args = fnArgsStr.isEmpty()
                ? Map.of()
                : objectMapper.readValue(fnArgsStr, Map.class);
        return new LlmResponse.ToolCall(
                item.get("name").asText(), item.get(CALL_ID_KEY).asText(), args);
    }

    private static void appendMessageText(JsonNode item, StringBuilder text) {
        JsonNode content = item.get(CONTENT_KEY);
        if (content == null || !content.isArray()) {
            return;
        }
        for (JsonNode part : content) {
            if ("output_text".equals(part.get("type").asText())
                    && part.has("text") && !part.get("text").isNull()) {
                text.append(part.get("text").asText());
            }
        }
    }

    @Override
    protected void addToolCallsToConversation(List<Map<String, Object>> conversation, LlmResponse response) {
        for (LlmResponse.ToolCall tc : response.toolCalls()) {
            String argsJson;
            try {
                argsJson = objectMapper.writeValueAsString(tc.arguments());
            } catch (Exception e) {
                argsJson = "{}";
            }
            conversation.add(Map.of(
                    "type", "function_call",
                    CALL_ID_KEY, tc.id(),
                    "name", tc.name(),
                    "arguments", argsJson
            ));
        }
    }

    @Override
    protected void addToolResultToConversation(List<Map<String, Object>> conversation,
                                                LlmResponse.ToolCall toolCall, String toolResult) {
        conversation.add(Map.of(
                "type", "function_call_output",
                CALL_ID_KEY, toolCall.id(),
                "output", toolResult
        ));
    }

    @Override
    protected void saveUsage(String rawResponse, long telegramUserId) {
        try {
            JsonNode jsonResponse = objectMapper.readTree(rawResponse);
            JsonNode usage = jsonResponse.get("usage");
            if (usage != null) {
                // The Responses API and Chat Completions name these differently.
                long promptTokens = firstPresent(usage, "input_tokens", "prompt_tokens");
                long completionTokens = firstPresent(usage, "output_tokens", "completion_tokens");
                long totalTokens = usage.has("total_tokens")
                        ? usage.get("total_tokens").asLong()
                        : promptTokens + completionTokens;
                tokenUsageRepository.save(new LlmTokenUsage(
                        telegramUserId, promptTokens, completionTokens, totalTokens));
            }
        } catch (Exception e) {
            log.warn("Failed to save token usage: {}", e.getMessage());
        }
    }

    /** Reads the first field present, so both API shapes are accepted. */
    private static long firstPresent(JsonNode usage, String... fieldNames) {
        for (String field : fieldNames) {
            if (usage.has(field)) {
                return usage.get(field).asLong();
            }
        }
        return 0;
    }

    @Override
    protected String getProviderName() {
        return "OpenAI";
    }

    private List<Map<String, Object>> buildToolDefinitions() {
        List<McpTool> tools = mcpRouter.listTools();
        List<Map<String, Object>> functions = new ArrayList<>();
        for (McpTool tool : tools) {
            Map<String, Object> fn = new LinkedHashMap<>();
            fn.put("type", "function");
            fn.put("name", tool.name());
            fn.put("description", tool.description());
            Map<String, Object> parameters = tool.inputSchema();
            if (parameters == null || parameters.isEmpty()) {
                parameters = Map.of("type", "object", "properties", Map.of());
            }
            fn.put("parameters", parameters);
            functions.add(fn);
        }
        return List.copyOf(functions);
    }
}

File: app/llm/openai_client.py
from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING, Any

import httpx

from app.config import Settings, reveal
from app.constants import SupportPrompt
from app.llm.base import (
    AbstractLlmClient,
    LlmProcessingException,
    LlmResponse,
    TokenUsage,
    ToolCall,
    is_balance_exhaustion_message,
)
from app.logging_config import TRACE
from app.logging_http import create_logging_hooks
from app.logging_redaction import safe_serialize
from app.retry import post_with_retry

if TYPE_CHECKING:
    from app.llm.mcp_router import McpRouter
    from app.rag.service import FaqEmbeddingService
    from app.storage.chat_history import ChatHistoryService
    from app.storage.database import DatabaseSessionManager

logger = logging.getLogger(__name__)

OPENAI_GPT_56_EFFORTS: frozenset[str] = frozenset({"none", "low", "medium", "high", "xhigh", "max"})
OPENAI_GPT_55_EFFORTS: frozenset[str] = frozenset({"none", "low", "medium", "high", "xhigh"})
OPENAI_GPT_55_PRO_EFFORTS: frozenset[str] = frozenset({"medium", "high", "xhigh"})
OPENAI_GPT_5_EFFORTS: frozenset[str] = frozenset({"minimal", "low", "medium", "high"})
OPENAI_GPT_5_PRO_EFFORTS: frozenset[str] = frozenset({"high"})
OPENAI_COMMON_REASONING_EFFORTS: frozenset[str] = frozenset({"low", "medium", "high"})


def supported_reasoning_efforts(model: str) -> frozenset[str] | None:
    """Return the documented effort set for a known OpenAI model family."""
    normalized = model.strip().lower()
    if normalized.startswith("gpt-5.6"):
        return OPENAI_GPT_56_EFFORTS
    if normalized.startswith("gpt-5.5-pro"):
        return OPENAI_GPT_55_PRO_EFFORTS
    if normalized.startswith("gpt-5.5"):
        return OPENAI_GPT_55_EFFORTS
    if normalized.startswith("gpt-5-pro"):
        return OPENAI_GPT_5_PRO_EFFORTS
    if normalized == "gpt-5" or normalized.startswith(("gpt-5-", "gpt-5-mini", "gpt-5-nano")):
        return OPENAI_GPT_5_EFFORTS
    if normalized.startswith(("gpt-5.1", "gpt-5.2", "gpt-5.3", "gpt-5.4")):
        return OPENAI_GPT_55_EFFORTS
    if normalized.startswith(("o1", "o3", "o4")):
        return OPENAI_COMMON_REASONING_EFFORTS
    return None


def supports_reasoning(model: str) -> bool:
    """Return whether an OpenAI model family accepts Responses reasoning config."""
    return supported_reasoning_efforts(model) is not None


class OpenAiClient(AbstractLlmClient):
    """Client for OpenAI Responses API (/responses) with tools and vision."""

    def __init__(
        self,
        settings: Settings,
        mcp_router: McpRouter,
        chat_history_service: ChatHistoryService,
        faq_embedding_service: FaqEmbeddingService,
        db_manager: DatabaseSessionManager | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        super().__init__(
            mcp_router=mcp_router,
            chat_history_service=chat_history_service,
            faq_embedding_service=faq_embedding_service,
            db_manager=db_manager,
        )
        api_key = reveal(settings.openai_api_key).strip()
        if not api_key:
            raise ValueError("OpenAI API key must not be null or blank")

        self.settings = settings
        self.api_key = api_key
        self.model = settings.openai_model or "gpt-5.6-luna"
        self.base_url = (settings.openai_base_url or "https://api.openai.com/v1").rstrip("/")
        self.temperature = settings.openai_temperature
        self.reasoning_effort = settings.reasoning_effort
        self.reasoning_supported = supports_reasoning(self.model)
        allowed_efforts = supported_reasoning_efforts(self.model)
        if allowed_efforts is not None and self.reasoning_effort not in allowed_efforts:
            raise ValueError(
                f"OpenAI model {self.model} не поддерживает REASONING_EFFORT={self.reasoning_effort}. "
                f"Допустимые значения: {', '.join(sorted(allowed_efforts))}"
            )
        self._http_client = http_client
        self._own_client = False
        self.tool_definitions = self._build_tool_definitions()
        self._log_reasoning_configuration()

    def _log_reasoning_configuration(self) -> None:
        if not self.reasoning_supported:
            if self.reasoning_effort != "none":
                logger.warning(
                    "OpenAI model %s does not support reasoning; "
                    "REASONING_EFFORT=%s is ignored and requests are sent without reasoning",
                    self.model,
                    self.reasoning_effort,
                )
            return
        logger.info(
            "OpenAI reasoning %s for model %s (REASONING_EFFORT=%s, MCP tools compatible)",
            "disabled" if self.reasoning_effort == "none" else "enabled",
            self.model,
            self.reasoning_effort,
        )
        if self.reasoning_effort != "none" and self.temperature is not None:
            logger.warning(
                "OPENAI_TEMPERATURE=%s is ignored while REASONING_EFFORT=%s is enabled",
                self.temperature,
                self.reasoning_effort,
            )

    @property
    def http_client(self) -> httpx.AsyncClient:
        if self._http_client is None:
            self._http_client = httpx.AsyncClient(
                timeout=httpx.Timeout(60.0), event_hooks=create_logging_hooks()
            )
            self._own_client = True
        return self._http_client

    async def close(self) -> None:
        if self._own_client and self._http_client is not None:
            await self._http_client.aclose()

    def supports_images(self) -> bool:
        return True

    def get_provider_name(self) -> str:
        return "OpenAI"

    def _build_tool_definitions(self) -> list[dict[str, Any]]:
        tools = self.mcp_router.list_tools()
        definitions: list[dict[str, Any]] = []
        for tool in tools:
            params = (
                tool.input_schema if tool.input_schema else {"type": "object", "properties": {}}
            )
            function: dict[str, Any] = {
                "type": "function",
                "name": tool.name,
                "description": tool.description or "",
                "parameters": params,
            }
            definitions.append(function)
        return definitions

    def build_request_body(self, conversation: list[dict[str, Any]]) -> dict[str, Any]:
        """Build Responses API request JSON body."""
        body: dict[str, Any] = {
            "model": self.model,
            "input": conversation,
        }
        if self.tool_definitions:
            body["tools"] = self.tool_definitions
            body["tool_choice"] = "auto"

        if self.reasoning_supported:
            body["reasoning"] = {"effort": self.reasoning_effort}

        if self.temperature is not None and (
            not self.reasoning_supported or self.reasoning_effort == "none"
        ):
            body["temperature"] = self.temperature

        return body

    def build_initial_conversation(
        self,
        user_message: str,
        telegram_user_id: int,
        faq_context: str | None = None,
        base64_image: str | None = None,
        mime_type: str | None = None,
        history: list[dict[str, Any]] | None = None,
    ) -> list[dict[str, Any]]:
        messages: list[dict[str, Any]] = []
        messages.append({"role": "system", "content": SupportPrompt.SYSTEM})

        dynamic_context = SupportPrompt.dynamic_context(faq_context, telegram_user_id)
        messages.append({"role": "system", "content": dynamic_context})

        if history:
            messages.extend(history)

        if base64_image and base64_image.strip():
            parts: list[dict[str, Any]] = []
            if user_message and user_message.strip():
                parts.append({"type": "input_text", "text": user_message})
            data_uri = f"data:{mime_type if mime_type else 'image/jpeg'};base64,{base64_image}"
            parts.append({"type": "input_image", "image_url": data_uri})
            messages.append({"role": "user", "content": parts})
        else:
            messages.append({"role": "user", "content": user_message})

        return messages

    async def call_api(
        self,
        conversation: list[dict[str, Any]],
        faq_context: str,
        telegram_user_id: int,
    ) -> dict[str, Any]:
        url = f"{self.base_url}/responses"
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        body = self.build_request_body(conversation)
        if logger.isEnabledFor(TRACE):
            effective_effort = self.reasoning_effort if self.reasoning_supported else "none"
            logger.log(
                TRACE,
                "OpenAI Responses API request (model=%s, configured_effort=%s, effective_effort=%s): %s",
                self.model,
                self.reasoning_effort,
                effective_effort,
                safe_serialize(body),
            )
        logger.debug(
            "OpenAI Responses API request (%d tools available)", len(self.tool_definitions)
        )

        response = await post_with_retry(
            self.http_client,
            url,
            headers=headers,
            json=body,
            description="OpenAI Responses API",
        )
        if response.status_code == 401:
            err_msg = f"OpenAI API error (model={self.model}, status=401)"
            logger.error(err_msg)
            raise LlmProcessingException(
                err_msg,
                "Произошла ошибка при обработке запроса. Попробуйте позже.",
                status_code=response.status_code,
            )

        if response.status_code >= 400:
            logger.error("OpenAI API error (model=%s, status=%d)", self.model, response.status_code)
            raise LlmProcessingException(
                f"OpenAI API error (model={self.model}, status={response.status_code})",
                "Произошла ошибка при обработке запроса. Попробуйте позже.",
                status_code=response.status_code,
                fallback_eligible=is_balance_exhaustion_message(response.text),
            )

        return self.decode_json(response)

    def parse_response(self, payload: dict[str, Any]) -> LlmResponse:
        try:
            output = payload.get("output")
            if not output or not isinstance(output, list):
                logger.error("No output array in OpenAI Responses API response: %s", payload)
                raise LlmProcessingException(
                    "Empty output",
                    "Не удалось получить ответ от модели. Попробуйте позже.",
                )

            text_builder: list[str] = []
            tool_calls: list[ToolCall] = []

            for item in output:
                item_type = item.get("type", "")
                if item_type == "function_call":
                    fn_name = item.get("name", "")
                    call_id = item.get("call_id", "")
                    fn_args_str = item.get("arguments", "{}")
                    if isinstance(fn_args_str, str):
                        try:
                            args = json.loads(fn_args_str) if fn_args_str else {}
                        except Exception:
                            args = {}
                    else:
                        args = fn_args_str or {}
                    tool_calls.append(ToolCall(name=fn_name, id=call_id, arguments=args))
                elif item_type == "message":
                    content = item.get("content", [])
                    if isinstance(content, list):
                        for part in content:
                            if part.get("type") == "output_text" and part.get("text"):
                                text_builder.append(part["text"])

            full_text = "".join(text_builder)
            if not full_text and not tool_calls:
                logger.warning("No text or tool calls in OpenAI response: %s", payload)
                raise LlmProcessingException(
                    "Empty response",
                    "Модель не вернула ответа. Попробуйте переформулировать вопрос.",
                )

            resp = LlmResponse(
                text=full_text,
                tool_calls=tool_calls,
                raw_parts=[item for item in output if isinstance(item, dict)],
            )
            if logger.isEnabledFor(TRACE):
                logger.log(
                    TRACE,
                    "OpenAI Responses API parsed response (model=%s): text=%s tool_calls=%s",
                    self.model,
                    full_text,
                    [
                        {"name": tc.name, "id": tc.id, "arguments": tc.arguments}
                        for tc in tool_calls
                    ],
                )
            return resp
        except LlmProcessingException:
            raise
        except Exception as e:
            logger.error("Failed to parse OpenAI response: %s", e)
            raise LlmProcessingException(
                f"Parse error: {e}", "Ошибка обработки ответа модели."
            ) from e

    def add_tool_calls_to_conversation(
        self,
        conversation: list[dict[str, Any]],
        response: LlmResponse,
    ) -> None:
        if response.raw_parts:
            conversation.extend(response.raw_parts)
            return
        for tc in response.tool_calls:
            conversation.append(
                {
                    "type": "function_call",
                    "call_id": tc.id,
                    "name": tc.name,
                    "arguments": json.dumps(tc.arguments or {}),
                }
            )

    def add_tool_result_to_conversation(
        self,
        conversation: list[dict[str, Any]],
        tool_call: ToolCall,
        tool_result: str,
    ) -> None:
        conversation.append(
            {
                "type": "function_call_output",
                "call_id": tool_call.id,
                "output": tool_result,
            }
        )

    def extract_usage(self, payload: dict[str, Any]) -> TokenUsage | None:
        usage = payload.get("usage")
        if not isinstance(usage, dict):
            return None
        prompt_tokens = int(usage.get("input_tokens") or usage.get("prompt_tokens") or 0)
        completion_tokens = int(usage.get("output_tokens") or usage.get("completion_tokens") or 0)
        return TokenUsage(
            prompt_tokens=prompt_tokens,
            completion_tokens=completion_tokens,
            total_tokens=int(usage.get("total_tokens") or prompt_tokens + completion_tokens),
        )
